### PR TITLE
Add initial version of select references flow

### PR DESCRIPTION
**Pull request by malcolmbaig**
## Context
We're implementing a feature that allows candidates to select which
references get submitted as part of their application (as opposed to
automatically using the first two that receive feedback). 

## Changes proposed in this pull request
- Add section complete functionality to references section
- Add reference selection functionality 

See commits for more detail.

Screenshot_20210520_162244

Screenshot_20210520_162435

## Guidance to review
In the review app:

- Ensure `reference_selection` feature active
- Get 2 or more references
- Mark section complete
- Select references (validations prevent you selecting greater or less than 2)

Note these two items aren't done yet and will follow in separate PRs (they're tracked on the Trello card):
- Marking the section incomplete when a selected reference is deleted
- Indicating selected references on the application form show page

## Link to Trello card
https://trello.com/c/hZVGCcxd

## Things to check

- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] API release notes have been updated if necessary
- [ ] This code does not rely on the addition/removal of Azure config environment variables in the same Pull Request
- [ ] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/forms/candidate_interface/reference/selection_form.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
+module CandidateInterface
+  class Reference::SelectionForm
+    include ActiveModel::Model
+    include ActiveModel::Validations::Callbacks
+
+    attr_accessor :application_form, :selected
+
+    before_validation :clean_up_checkbox_params
```
Comment (davidgisbey) [inline app/forms/candidate_interface/reference/selection_form.rb:8]: We should rework our forms which manually cast booleans to use this 👌

File: app/forms/candidate_interface/reference/selection_form.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    def available_references
+      application_form.application_references.feedback_provided
+    end
+
+    def save!
+      return false unless valid?
+
+      available_references.update_all(selected: false)
+      available_references.where(id: selected).update_all(selected: true)
```
Comment (davidgisbey) [inline app/forms/candidate_interface/reference/selection_form.rb:19]: Can we just do 

```
available_references.where.not(id: selected).update_all(selected: false) 
available_references.where(id: selected).update_all(selected: true)
```

File: app/views/candidate_interface/references/selection/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<% content_for :title, t('page_titles.references') %>
+<% content_for :before_content, govuk_back_link_to(candidate_interface_application_form_path, 'Back to application') %>
+
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
+    <h1 class="govuk-heading-xl"> <%= t('page_titles.references_select') %> </h1>
+
+    <%= form_with(model: @selection_form, url: candidate_interface_select_references_path, method: :post) do |f| %>
```
Comment (davidgisbey) [inline app/views/candidate_interface/references/selection/new.html.erb:9]: should this be a :patch

File: app/controllers/candidate_interface/references/review_controller.rb [hunk 4]
```diff
@@ -22,6 +22,11 @@ def complete
           render 'show_for_reference_selection' and return
         end
 
+        if !current_application.minimum_references_available_for_selection?
```
Comment (davidgisbey) [inline app/controllers/candidate_interface/references/review_controller.rb:25]: I'm wondering if we just shouldn't show the radios in this situation and have a continue button. Wdyt?

Comment (malcolmbaig) [reply app/controllers/candidate_interface/references/review_controller.rb:25]: @JR-G is implementing the design improvements suggested by @paulrobertlloyd, which result in this moving elsewhere. Let's revisit in https://github.com/DFE-Digital/apply-for-teacher-training/pull/4824 when that's ready for review!

**Review by davidgisbey — APPROVED** [review]
LGTM once the linter stops complaining 👍 Great work

**Pull request merged**